File: sleep-viz/src/lib/formatters.ts
import { format } from 'date-fns'

export function formatDuration(minutes: number): string {
  const hrs = Math.floor(minutes / 60)
  const mins = Math.round(minutes % 60)
  if (hrs === 0) return `${mins}m`
  return mins === 0 ? `${hrs}h` : `${hrs}h ${mins}m`
}

export function formatTime(date: Date): string {
  return format(date, 'h:mm a')
}

export function formatNightDate(nightDate: string): string {
  return format(new Date(nightDate + 'T12:00:00'), 'EEE, MMM d')
}

export function formatPercent(value: number): string {
  return `${Math.round(value)}%`
}

export function formatMinutesAsTime(minutesFromMidnight: number): string {
  let mins = minutesFromMidnight
  if (mins < 0) mins += 24 * 60
  const hrs = Math.floor(mins / 60) % 24
  const m = Math.round(mins % 60)
  const period = hrs >= 12 ? 'PM' : 'AM'
  const h = hrs % 12 || 12
  return `${h}:${m.toString().padStart(2, '0')} ${period}`
}

export function formatBpm(value: number | null): string {
  if (value === null) return '—'
  return `${Math.round(value)} bpm`
}

export function formatMs(value: number | null): string {
  if (value === null) return '—'
  return `${Math.round(value)} ms`
}

File: Amir-SleepApp/Amir-SleepApp/Utilities/Formatters.swift
import Foundation

func formatDuration(minutes: Double) -> String {
    let hrs = Int(minutes) / 60
    let mins = Int(minutes.rounded()) % 60
    if hrs == 0 { return "\(mins)m" }
    return mins == 0 ? "\(hrs)h" : "\(hrs)h \(mins)m"
}

func formatTime(_ date: Date) -> String {
    let formatter = DateFormatter()
    formatter.dateFormat = "h:mm a"
    return formatter.string(from: date)
}

func formatNightDate(_ nightDate: String) -> String {
    let inputFormatter = DateFormatter()
    inputFormatter.dateFormat = "yyyy-MM-dd"
    guard let date = inputFormatter.date(from: nightDate) else { return nightDate }
    let outputFormatter = DateFormatter()
    outputFormatter.dateFormat = "EEE, MMM d"
    return outputFormatter.string(from: date)
}

func formatPercent(_ value: Double) -> String {
    "\(Int(value.rounded()))%"
}

func formatBpm(_ value: Double?) -> String {
    guard let value else { return "—" }
    return "\(Int(value.rounded())) bpm"
}

func formatMs(_ value: Double?) -> String {
    guard let value else { return "—" }
    return "\(Int(value.rounded())) ms"
}

func formatMinutesAsTime(_ minutesFromMidnight: Double) -> String {
    var mins = minutesFromMidnight
    if mins < 0 { mins += 24 * 60 }
    let hrs = Int(mins) / 60 % 24
    let m = Int(mins.rounded()) % 60
    let period = hrs >= 12 ? "PM" : "AM"
    let h = hrs % 12 == 0 ? 12 : hrs % 12
    return String(format: "%d:%02d %@", h, m, period)
}
